src/mod_amp.erl

```diff
@@ -29,32 +33,40 @@
 
 -spec start(mongooseim:host_type(), gen_mod:module_opts()) -> ok.
 start(HostType, _Opts) ->
-    ejabberd_hooks:add(hooks(HostType)).
+    ejabberd_hooks:add(hooks(HostType)),
+    gen_hook:add_handlers(c2s_hooks(HostType)).
 
 -spec stop(mongooseim:host_type()) -> ok.
 stop(HostType) ->
-    ejabberd_hooks:delete(hooks(HostType)).
+    ejabberd_hooks:delete(hooks(HostType)),
+    gen_hook:delete_handlers(c2s_hooks(HostType)).
 
 -spec supported_features() -> [atom()].
 supported_features() -> [dynamic_domains].
 
+-spec hooks(mongooseim:host_type()) -> [ejabberd_hooks:hook()].
 hooks(HostType) ->
     [{c2s_stream_features, HostType, ?MODULE, c2s_stream_features, 50},
      {disco_local_features, HostType, ?MODULE, disco_local_features, 99},
-     {c2s_preprocessing_hook, HostType, ?MODULE, run_initial_check, 10},
      {amp_verify_support, HostType, ?AMP_RESOLVER, verify_support, 10},
      {amp_check_condition, HostType, ?AMP_RESOLVER, check_condition, 10},
-     {amp_determine_strategy, HostType, ?AMP_STRATEGY, determine_strategy, 10}].
+     {amp_determine_strategy, HostType, ?AMP_STRATEGY, determine_strategy, 10},
+     {xmpp_send_element, HostType, ?MODULE, xmpp_send_element, 10}].
+
+-spec c2s_hooks(mongooseim:host_type()) -> gen_hook:hook_list(mongoose_c2s_hooks:hook_fn()).
+c2s_hooks(HostType) ->
+    [{c2s_preprocessing_hook, HostType, fun ?MODULE:run_initial_check/3, #{}, 10}].
 
 %% API
 
--spec run_initial_check(mongoose_acc:t(), ejabberd_c2s:state()) -> mongoose_acc:t().
-run_initial_check(#{result := drop} = Acc, _C2SState) ->
-    Acc;
-run_initial_check(Acc, _C2SState) ->
+-spec run_initial_check(mongoose_acc:t(), mongoose_c2s:hook_params(), gen_hook:extra()) ->
+    mongoose_c2s_hooks:hook_result().
+run_initial_check(#{result := drop} = Acc, _Params, _Extra) ->
```

Comment:
Actually this is a bug in the old code, the acc will store the result in something like `#{{hook, result} := drop}`, in `mongoose_acc` keys are pairs with a namespace. We shouldn't do that anyway, `mongoose_acc` is an abstract data structure, we should do `case mongoose_acc:get(hook, result, undefined, Acc) of` and match on `drop` or on whatever else 🤔